### PR TITLE
Extract GlobalSnapshot to shared

### PR DESCRIPTION
**Pull request by kpudlik**

### CODE REVIEW

File: clean-bsp.sh [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env sh
```
Comment (marcinwadon) [inline clean-bsp.sh]: should it be kept in the repository? 🤔

File: modules/core/src/main/scala/org/tessellation/domain/snapshot/SnapshotStorage.scala [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package org.tessellation.domain.snapshot
+
+import org.tessellation.schema.snapshot.Snapshot
+import org.tessellation.schema.transaction.Transaction
+import org.tessellation.schema.{Block, SnapshotOrdinal}
+import org.tessellation.security.hash.Hash
+import org.tessellation.security.signature.Signed
+
+trait SnapshotStorage[F[_], T <: Transaction, B <: Block[T], S <: Snapshot[T, B]] {
```
Comment (marcinwadon) [inline modules/core/src/main/scala/org/tessellation/domain/snapshot/SnapshotStorage.scala:9]: what happened with `Global` prefix? 🤔

File: modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotConsensusFunctions.scala [hunk 3]
```diff
@@ -164,7 +164,7 @@ object GlobalSnapshotConsensusFunctions {
 
         returnedDAGEvents = getReturnedDAGEvents(acceptanceResult)
 
-        globalSnapshot = GlobalSnapshot(
+        globalSnapshot = schema.GlobalSnapshot(
```
Comment (marcinwadon) [inline modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotConsensusFunctions.scala:167]: typo

File: modules/dag-l1/src/test/scala/org/tessellation/dag/l1/domain/snapshot/programs/SnapshotProcessorSuite.scala [hunk 4]
```diff
@@ -141,7 +141,7 @@ object SnapshotProcessorSuite extends SimpleIOSuite with TransactionGenerator {
     )
 
   def generateSnapshot(peerId: PeerId): GlobalSnapshot =
-    GlobalSnapshot(
+    schema.GlobalSnapshot(
```
Comment (marcinwadon) [inline modules/dag-l1/src/test/scala/org/tessellation/dag/l1/domain/snapshot/programs/SnapshotProcessorSuite.scala:144]: typo

File: modules/dag-shared/src/main/scala/org/tessellation/dag/package.scala [hunk 5]
```diff
@@ -20,8 +20,8 @@ package object dag {
     classOf[DAGBlock] -> 603,
     DAGBlock.OrderingInstance.getClass -> 604,
     classOf[GlobalSnapshotInfo] -> 606,
-    classOf[DAGBlockAsActiveTip] -> 610,
-    DAGBlockAsActiveTip.OrderingInstance.getClass -> 611,
+    classOf[BlockAsActiveTip[_]] -> 610,
+//    BlockAsActiveTip.OrderingInstance.getClass -> 611,
```
Comment (marcinwadon) [inline modules/dag-shared/src/main/scala/org/tessellation/dag/package.scala:24]: ?

File: modules/core/src/test/scala/org/tessellation/infrastructure/snapshot/SnapshotStorageSuite$.scala [hunk 6]
```diff
@@ -29,11 +28,11 @@ import fs2.io.file.Path
 import weaver.MutableIOSuite
 import weaver.scalacheck.Checkers
 
-object GlobalSnapshotStorageSuite extends MutableIOSuite with Checkers {
+object SnapshotStorageSuite$ extends MutableIOSuite with Checkers {
```
Comment (marcinwadon) [inline modules/core/src/test/scala/org/tessellation/infrastructure/snapshot/SnapshotStorageSuite$.scala:31]: why `$`?

**Review by marcinwadon — CHANGES_REQUESTED** [review]

File: modules/dag-shared/src/main/scala/org/tessellation/dag/package.scala (hunk 5 above)
Comment (kpudlik) [inline modules/dag-shared/src/main/scala/org/tessellation/dag/package.scala:24]: I am wondering if we can safely get rid of it. I was forced to implement `Ordering[BlockAsActiveTip]` as `def` so it works with any block but I am worried about backward compatibility 🤔

File: clean-bsp.sh (hunk 1 above)
Comment (kpudlik) [reply clean-bsp.sh]: pushed by accident, fixed

File: modules/core/src/main/scala/org/tessellation/domain/snapshot/SnapshotStorage.scala (hunk 2 above)
Comment (kpudlik) [reply modules/core/src/main/scala/org/tessellation/domain/snapshot/SnapshotStorage.scala:9]: We do not need `Global` prefix anymore. The storage works the same for `GlobalSnapshot` and `CurrencySnapshot` thus we can just parametrize it by `S`

File: modules/dag-shared/src/main/scala/org/tessellation/dag/package.scala (hunk 5 above)
Comment (kpudlik) [reply modules/dag-shared/src/main/scala/org/tessellation/dag/package.scala:24]: Read my comment below that line on GH

File: modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotConsensusFunctions.scala (hunk 3 above)
Comment (kpudlik) [reply modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotConsensusFunctions.scala:167]: Just IntelliJ trying to be smart

File: modules/dag-l1/src/test/scala/org/tessellation/dag/l1/domain/snapshot/programs/SnapshotProcessorSuite.scala (hunk 4 above)
Comment (kpudlik) [reply modules/dag-l1/src/test/scala/org/tessellation/dag/l1/domain/snapshot/programs/SnapshotProcessorSuite.scala:144]: Just IntelliJ trying to be smart

File: modules/core/src/main/scala/org/tessellation/domain/snapshot/SnapshotStorage.scala (hunk 2 above)
Comment (kpudlik) [reply modules/core/src/main/scala/org/tessellation/domain/snapshot/SnapshotStorage.scala:9]: Eventually we can make a type alias for `Global` one

File: modules/dag-shared/src/main/scala/org/tessellation/dag/package.scala (hunk 5 above)
Comment (kpudlik) [reply modules/dag-shared/src/main/scala/org/tessellation/dag/package.scala:24]: @marcinwadon ^

Comment (kpudlik) [reply modules/dag-shared/src/main/scala/org/tessellation/dag/package.scala:24]: https://github.com/Constellation-Labs/tessellation/pull/546/files#diff-1d5cf500b56589e3429a0234cad7a312853d5e5972487c0eb1d783a682310b83

Comment (kpudlik) [reply modules/dag-shared/src/main/scala/org/tessellation/dag/package.scala]: Ok, I have confirmed that I need to keep it

File: modules/core/src/main/scala/org/tessellation/domain/snapshot/SnapshotStorage.scala (hunk 2 above)
Comment (marcinwadon) [reply modules/core/src/main/scala/org/tessellation/domain/snapshot/SnapshotStorage.scala:9]: ok, I'm fine with this :+1:

Comment (marcinwadon) [inline modules/core/src/main/scala/org/tessellation/domain/snapshot/SnapshotStorage.scala:9]: do you need to specify `T` and `B`? looks unused 🤔

Comment (kpudlik) [reply modules/core/src/main/scala/org/tessellation/domain/snapshot/SnapshotStorage.scala:9]: I can use wildcard `S <: Snapshot[_, _]`

File: modules/shared/src/main/scala/org/tessellation/schema/BlockAsActiveTip.scala [hunk 7]
```diff
@@ -1,10 +1,27 @@
...
+  usageCount: NonNegLong
+)
+
+object BlockAsActiveTip {
+  implicit def order[B <: Block[_]: Order]: Order[BlockAsActiveTip[B]] = Order.whenEqual(
+    Order.by[BlockAsActiveTip[B], Signed[B]](_.block),
+    Order.by[BlockAsActiveTip[B], NonNegLong](_.usageCount)
+  )
+
+  implicit object OrderingInstance extends OrderBasedOrdering[BlockAsActiveTip[DAGBlock]]
```
Comment (marcinwadon) [inline modules/shared/src/main/scala/org/tessellation/schema/BlockAsActiveTip.scala:26]: hm, looks like mixed abstraction with implementation (`DAGBlock` is a specific implementation)

Comment (kpudlik) [reply modules/shared/src/main/scala/org/tessellation/schema/BlockAsActiveTip.scala:26]: I do not like it either. I think we can just keep it next to `DAGBlock`.

File: modules/dag-shared/src/main/scala/org/tessellation/dag/package.scala (hunk 5 above)
Comment (TheMMaciek) [reply modules/dag-shared/src/main/scala/org/tessellation/dag/package.scala]: Yeah I think we can register both, e.i. ordering instance for `BlockAsActiveTip` parametrized by `DAGBlock` and we will register another ordering instance for `BlockAsActiveTip` parametrized by `CurrencyBlock`. Would that work?

File: modules/core/src/main/scala/org/tessellation/domain/snapshot/SnapshotStorage.scala (hunk 2 above)
Comment (TheMMaciek) [reply modules/core/src/main/scala/org/tessellation/domain/snapshot/SnapshotStorage.scala]: Given that we are not fetching blocks or transactions (unwrapped from within the snapshot) from this storage, the wildcards should work.

Comment (marcinwadon) [reply modules/core/src/main/scala/org/tessellation/domain/snapshot/SnapshotStorage.scala]: then probably we need it parametrized @kpudlik

File: modules/core/src/test/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotSerializationSuite.scala [hunk 8]
```diff
@@ -6,11 +6,10 @@ import cats.effect.{IO, Resource}
 import cats.syntax.all._
 
 import org.tessellation.dag.dagSharedKryoRegistrar
-import org.tessellation.dag.snapshot.GlobalSnapshot
 import org.tessellation.ext.crypto._
 import org.tessellation.ext.kryo._
 import org.tessellation.kryo.KryoSerializer
-import org.tessellation.schema._
+import org.tessellation.schema.{GlobalSnapshot, _}
```
Comment (TheMMaciek) [inline modules/core/src/test/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotSerializationSuite.scala:12]: `_` should be enough

**Review by TheMMaciek — APPROVED** [review]
Looks good overall

File: modules/dag-shared/src/main/scala/org/tessellation/dag/package.scala (hunk 5 above)
Comment (kpudlik) [reply modules/dag-shared/src/main/scala/org/tessellation/dag/package.scala]: Ok I left both for now. We will see how it works when then Currency is implemented

**Review by marcinwadon — APPROVED** [review]

**Review by marcinwadon — APPROVED** [review]

**Review by TheMMaciek — APPROVED** [review]

**Pull request merged**